utils.py

```diff
@@ -724,6 +725,111 @@ def get_hashable_value(value):
+
+    Returns:
+        str. The converted HTML string.
+    """
+    if not len(html_data):
+        return html_data
+
+    soup = bs4.BeautifulSoup(html_data, 'html.parser')
+
+    allowed_tag_list = (
+        feconf.RTE_CONTENT_SPEC[
+            'RTE_TYPE_TEXTANGULAR']['ALLOWED_TAG_LIST'])
+    allowed_parent_list = (
+        feconf.RTE_CONTENT_SPEC[
+            'RTE_TYPE_TEXTANGULAR']['ALLOWED_PARENT_LIST'])
+
+    # To remove all tags except those in allowed tag list.
+    all_tags = soup.findAll()
+    for tag in all_tags:
+        if tag.name == 'td' and tag.next_sibling:
+            tag.string = tag.string + " "
+        if tag.name == 'div' or tag.name == 'tr':
+            tag.name = 'p'
+        elif tag.name not in allowed_tag_list:
+            tag.unwrap()
+
+    # Removal of tags can break the soup into parts which are continuous
+    # and not wrapped in any tag. This part recombines the continuous
+    # parts not wrapped in any tag.
+    soup = bs4.BeautifulSoup(str(soup), 'html.parser')
+
+    oppia_inline_components = [
+        'oppia-noninteractive-link', 'oppia-noninteractive-math']
+    oppia_block_components = [
+        'oppia-noninteractive-image',
+        'oppia-noninteractive-video',
+        'oppia-noninteractive-collapsible',
+        'oppia-noninteractive-tabs'
+    ]
+
+    # Ensure that every content in html is wrapped in a tag.
+    for content in soup.contents:
+        if not content.name:
+            content.wrap(soup.new_tag('p'))
+
+    # Ensure that every line break is a child of any of its allowed parents.
+    enforce_valid_parent(soup, 'br', allowed_parent_list['br'], 'p')
+
+    # Ensure that every oppia inline component is a child of any
+    # of its allowed parents.
+    for tag_name in oppia_inline_components:
+        enforce_valid_parent(soup, tag_name, allowed_parent_list[tag_name], 'p')
+
+    # Ensure that every oppia block component is a child of any of its
+    # allowed parents.
+    for tag_name in oppia_block_components:
+        enforce_valid_parent(soup, tag_name, allowed_parent_list[tag_name], 'p')
+
+    # Ensure that every bold tag is a child of any of its allowed parents.
+    enforce_valid_parent(soup, 'b', allowed_parent_list['b'], 'p')
+
+    # Ensure that every italics tag is a child of any of its allowed parents.
+    enforce_valid_parent(soup, 'i', allowed_parent_list['i'], 'p')
+
+    # Ensure that p tag has a valid parent. p tags having parent tag as p
+    # is checked separately since in that case the child p tag is to
+    # be unwrapped instead of the parent p tag.
+    for p in soup.findAll('p'):
+        if p.parent.name != 'p' and (
+                p.parent.name not in allowed_parent_list['p']):
+            p.parent.unwrap()
+
+    # Ensure that p tag is not wrapped in p tag.
+    for p in soup.findAll('p'):
+        if p.parent.name == 'p':
+            p.unwrap()
+
+    # Keeping all line break tags as <br>.
+    return str(soup).replace('<br/>', '<br>')
```

Comment:
You have beautifulsoup to validate the HTML. Why not use that to identify nodes of type 'br', rather than treating the HTML as a string?

Reply:
Beautiful soup automatically changes `<br>` to `<br/>`, so it had to be replaced directly in the string. Also, when any html string with `<br/>` is stored in exploration html strings they are stored as `<br>`. Since both of these should match and `<br>` and `<br/>` have same working, so I replaced the tag in html data during migration.

Reply:
That's a good explanation. Please add a docstring to say exactly that, the context is important here as to why we're operating on the raw strings.